lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+        _updatePaused(false);
+        _reportedException = null;
+        break;
+      case EventKind.kPauseStart:
+      case EventKind.kPauseExit:
+      case EventKind.kPauseBreakpoint:
+      case EventKind.kPauseInterrupted:
+      case EventKind.kPauseException:
+      case EventKind.kPausePostRequest:
+        _reportedException = event.exception;
+        _updatePaused(true);
+        break;
+      case EventKind.kBreakpointAdded:
+        _breakpoints.value.add(event.breakpoint);
+        _breakpoints.add(_breakpoints.value);
+        break;
+      case EventKind.kBreakpointResolved:
+        _breakpoints.value.remove(event.breakpoint);
+        _breakpoints.value.add(event.breakpoint);
+        _breakpoints.add(_breakpoints.value);
+        break;
+      case EventKind.kBreakpointRemoved:
+        _breakpoints.value.remove(event.breakpoint);
+        _breakpoints.add(_breakpoints.value);
+        break;
+    }
+  }
+
+  void _clearCaches() {
+    _scriptCache.clear();
+    _lastEvent = null;
+    _reportedException = null;
+  }
+
+  void dispose() {
+    _debugSubscription?.cancel();
+  }
+
+  void _updatePaused(bool value) {
+    if (_paused.value != value) {
+      _paused.add(value);
+    }
+  }
+
+  Future<Script> getScript(ScriptRef scriptRef) async {
+    if (!_scriptCache.containsKey(scriptRef.id)) {
+      _scriptCache[scriptRef.id] =
+          await service.getObject(isolateRef.id, scriptRef.id);
+    }
+
+    return _scriptCache[scriptRef.id];
+  }
+
+  Pos calculatePosition(Script script, int tokenPos) {
+    final List<List<int>> table = script.tokenPosTable;
+    if (table == null) {
+      return null;
+    }
+
+    for (List<int> row in table) {
+      if (row == null || row.isEmpty) {
+        continue;
+      }
+      final int line = row.elementAt(0);
+      int index = 1;
+
+      while (index < row.length - 1) {
+        if (row.elementAt(index) == tokenPos) {
+          return new Pos(line, row.elementAt(index + 1));
+        }
+        index += 2;
+      }
+    }
+
+    return null;
+  }
+
+  String commonScriptPrefix;
+  LibraryRef rootLib;
+
+  void setRootLib(LibraryRef rootLib) {
+    this.rootLib = rootLib;
+
+    String scriptPrefix = rootLib.uri;
+    if (scriptPrefix.startsWith('package:')) {
+      scriptPrefix = scriptPrefix.substring(0, scriptPrefix.indexOf('/') + 1);
+    } else if (scriptPrefix.contains('/lib/')) {
+      scriptPrefix =
+          scriptPrefix.substring(0, scriptPrefix.lastIndexOf('/lib/'));
+      if (scriptPrefix.contains('/')) {
+        scriptPrefix =
+            scriptPrefix.substring(0, scriptPrefix.lastIndexOf('/') + 1);
+      }
+    } else if (scriptPrefix.contains('/bin/')) {
+      scriptPrefix =
+          scriptPrefix.substring(0, scriptPrefix.lastIndexOf('/bin/'));
+      if (scriptPrefix.contains('/')) {
+        scriptPrefix =
+            scriptPrefix.substring(0, scriptPrefix.lastIndexOf('/') + 1);
+      }
+    } else if (scriptPrefix.contains('/example/')) {
```

Comment:
is this hard coding of /example/ needed? Seems a bit surprising. Generally it would be helpful to move this logic for determining the root directory to a shared util as I also need it for the inspector to clarify what directories are considered the local project for highlighting only user created widgets.